## Single-cell transcriptomic, epigenomic and spatial metabolomic profiling redefines the human kidney anatomy

This repository documents the scripts to generate data for our manuscript studying human kidney anatomical regions with SHARE-seq and imaging mass spectrometry. <link> <br>
More information about MALDIpy, a package for single-cell analysis of MALDI-MS imaging mass spectrometry data, can be found in another GitHub repository: https://github.com/TheHumphreysLab/MALDIpy. <br>

The raw (.fastq) and pre-processed (count matrix) data and metadata have been deposited in NCBIâ€™s Gene Expression Omnibus and are available through GEO Series accession number .  <br> <br>
A searchable database is available at our Kidney Interactive Transcriptomics (K.I.T.) website: http://humphreyslab.com/SingleCell/.<link> <br><br>
Pre-processing of raw fastq files was performed as previously described (Ma et al. 2020, Cell): https://github.com/masai1116/SHARE-seq-alignment and https://github.com/masai1116/SHARE-seq-alignmentV2/.<br>

### Descriptions

#### 1. Scripts for Figure 1<br>
Quality control and single-cell clustering<br>
Pseudobulk trajectory ordering (Fig. 1c)<br>
Single-cell visualization (Fig. 1d)<br>
Gene expression visualization (Fig. 1e)<br>


#### 2. Scripts for Figure 2<br>
Single-cell visualization (Fig. 2a)<br>
Gene expression visualization (Fig. 2b)<br>
Single-cell TF activity analysis (Fig. 2c)<br>
Single-cell pathway activity analysis (Fig. 2d)<br>
Single-cell trajectory ordering (Fig. 2f)<br>
Fate prediction analysis with time-series scRNAseq data (Fig. 2g)<br>

#### 3. Scripts for Figure 3<br> 
Quality control and single-cell subclustering<br>
Single-cell visualization (Fig. 6a/b)<br>
Gene expression visualization (Fig. 6c/d/f)<br>

#### 4. Scripts for Figure 4<br>
Quality control and single-cell subclustering<br>
Single-cell visualization (Fig. 7a/b)<br>
Gene module scoring (Fig. 7e)<br>
Commands used for CellPhoneDB analysis<br>
Statistical analysis (Fig. 7f/g)<br>

#### 5. Scripts for Figure 5<br>
RT barcode - Sample Lookup Table<br>
Meta info of all cells used for clustering<br>

#### 6. Scripts for Figure 6&7<br>
RT barcode - Sample Lookup Table<br>
Meta info of all cells used for clustering<br>

#### 7. Scripts for other multimodal analysis (multiome_code)<br>
1. Weighted Nearest Neighbour (WNN)
2. chromVAR
3. Cicero
4. RENIN


<br>
**************<br>


Find us on Twitter: 
<br/>
  <a href="https://twitter.com/HumphreysLab?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false"> @HumphreysLab</a>
<br/><br/>
